deltacat/compute/compactor_v2/steps/merge.py

```diff
@@ -257,14 +257,170 @@ def _copy_all_manifest_files_from_old_hash_buckets(
     return materialize_result_list
 
 
+def _does_hash_bucket_idx_have_compacted_table(input: MergeInput, hb_idx: int) -> bool:
+    return (
+        input.round_completion_info
+        and input.round_completion_info.hb_index_to_entry_range
+        and input.round_completion_info.hb_index_to_entry_range.get(str(hb_idx))
+        is not None
+    )
+
+
+def _flatten_dfe_list(
+    df_envelopes_list: List[List[DeltaFileEnvelope]],
+) -> List[DeltaFileEnvelope]:
+    if not df_envelopes_list:
+        return []
+    return [d for dfe_list in df_envelopes_list for d in dfe_list]
+
+
+def _sort_df_envelopes(
+    df_envelopes: List[DeltaFileEnvelope],
+    key: Callable = lambda df: (df.stream_position, df.file_index),
+) -> List[DeltaFileEnvelope]:
+    if not df_envelopes:
+        return []
+    return sorted(
+        df_envelopes,
+        key=key,
+        reverse=False,  # ascending
+    )
+
+
+def group_by_upsert_delete_sequence(
+    df_envelopes: List[DeltaFileEnvelope],
+) -> Iterator[Tuple[List, List]]:
+    """
+    TODO
+    """
+    iter_df_envelopes = iter(df_envelopes)
+    upserts, deletes = [], []
+    for is_upsert, group_envelopes in itertools.groupby(
+        iter_df_envelopes, lambda x: x.delta_type == DeltaType.UPSERT
+    ):
+        if is_upsert:
+            upserts.extend(group_envelopes)
+        else:
+            deletes.extend(group_envelopes)
+        if deletes or not (deletes or upserts):
+            yield upserts, deletes
+            upserts, deletes = [], []
+    if upserts:
+        yield upserts, []
+
+
+def _compact_table_v2(
+    input: MergeInput,
+    dfe_list: List[List[DeltaFileEnvelope]],
+    hb_idx: int,
+    has_deletes: bool,
+) -> Tuple[pa.Table, int, int]:
+    df_envelopes: List[DeltaFileEnvelope] = _sort_df_envelopes(
+        _flatten_dfe_list(dfe_list)
+    )
+    delete_file_envelopes = input.delete_file_envelopes or []
+    all_dfes: List[DeltaFileEnvelope] = _sort_df_envelopes(
+        delete_file_envelopes + df_envelopes
+    )
+    prev_table, table = None, None
+    total_incremental_len = 0
+    total_deduped_records = 0
+    total_dropped_records = 0
+    if _does_hash_bucket_idx_have_compacted_table(input, hb_idx):
+        table = _download_compacted_table(
+            hb_index=hb_idx,
+            rcf=input.round_completion_info,
+            read_kwargs_provider=input.read_kwargs_provider,
+            deltacat_storage=input.deltacat_storage,
+            deltacat_storage_kwargs=input.deltacat_storage_kwargs,
+        )
+        if has_deletes:
+            table, partial_dropped_rows = input.delete_strategy.apply_all_deletes(
+                table, input.delete_file_envelopes
+            )
+            total_dropped_records += partial_dropped_rows
+    prev_table = table
+    for i, (upsert_group, delete_group) in enumerate(
+        group_by_upsert_delete_sequence(all_dfes)
+    ):
```

Comment:
This is complicated. You don't need to process `upsert_group` and `delete_group` at once. The former requires you to call `_apply_upserts` and latter requires you to call `apply_all_deletes`. At the end of the day, you have an arrow table as result.